### PR TITLE
core/muxing: Drop `Sync` requirement for `StreamMuxer` on `StreamMuxerBox`

### PR DESCRIPTION
**Pull request by thomaseizinger**
# Description

`StreamMuxerBox` is never shared across threads but owned by a single
connection. Restricting it to be `Sync` unnecessarily limits the design
space around the `StreamMuxer` trait and its implementations.

## Links to any relevant issues

Required for https://github.com/libp2p/rust-libp2p/pull/2765.

## Change checklist

- [x] I have performed a self-review of my own code
~- [ ] I have made corresponding changes to the documentation~
~- [ ] I have added tests that prove my fix is effective or that my feature works~
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: core/CHANGELOG.md [hunk 1]
```diff
@@ -3,9 +3,11 @@
 - Remove `StreamMuxer::poll_event` in favor of individual functions: `poll_inbound`, `poll_outbound`
   and `poll_address_change`. Consequently, `StreamMuxerEvent` is also removed. See [PR 2724].
 - Drop `Unpin` requirement from `SubstreamBox`. See [PR 2762].
+- Drop `Sync` requirement on `StreamMuxer` for constructing `StreamMuxerBox`. See [PR XXXX].
 
 [PR 2724]: https://github.com/libp2p/rust-libp2p/pull/2724
 [PR 2762]: https://github.com/libp2p/rust-libp2p/pull/2762
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```
Comment (thomaseizinger) [inline core/CHANGELOG.md]:
```suggestion
- Drop `Sync` requirement on `StreamMuxer` for constructing `StreamMuxerBox`. See [PR 2775].

[PR 2724]: https://github.com/libp2p/rust-libp2p/pull/2724
[PR 2762]: https://github.com/libp2p/rust-libp2p/pull/2762
[PR 2775]: https://github.com/libp2p/rust-libp2p/pull/2775
```

File: core/src/transport/upgrade.rs [hunk 2]
```diff
@@ -299,7 +299,7 @@ impl<T> Multiplexed<T> {
         T::Dial: Send + 'static,
         T::ListenerUpgrade: Send + 'static,
         T::Error: Send + Sync,
-        M: StreamMuxer + Send + Sync + 'static,
+        M: StreamMuxer + Send + 'static,
         M::Substream: Send + Unpin + 'static,
         M::Error: Send + Sync + 'static,
```
Comment (elenaf9) [inline core/src/transport/upgrade.rs]: Could we also drop the other `Sync` requirements here?

**Review by elenaf9 — APPROVED** [review]

File: core/src/transport/upgrade.rs (hunk 2 above)
Comment (thomaseizinger) [reply core/src/transport/upgrade.rs]: Maybe. I'd say we do it in a different PR though :)

**Review by mxinden — APPROVED** [review]
My favorite trait bound to be dropped. :pray:

**Pull request merged**